### PR TITLE
Upgrade Qiskit to 1.0.2

### LINKED ISSUE #2208: Any qiskit circuit that requires conversion to qasm breaks as of  qiskit 1.0.1
## Issue Description
 
Example: Take the cdr case. We want to generate clifford training data with the input as qiskit quantum circuit. This fails with the error as
`'QuantumCircuit' object has no attribute 'qasm'`

## How to Reproduce
Minimum steps
```
from qiskit import QuantumCircuit
from mitiq.cdr import generate_training_circuits
qc = QuantumCircuit(3)
qc.h(0)
qc.cx(0,1)
qc.cx(1,2)

cd = generate_training_circuits(qc, 10, 0.1)
```

This generates following error
``` return from_qasm(circuit.qasm())

AttributeError: 'QuantumCircuit' object has no attribute 'qasm'
```

## Root Cause
QuantumCircuit object from qiskit 1.0.1 no longer has qasm as an attribute

## Possible solutions and workarounds
Fix - Replace circuit.qasm() in from_qiskit function with qiskit.qasm2.dumps(circuit)
This fix also works with qiskit 0.45 
---

<details>
<summary>Environment Context</summary>

Use the `about()` function to summarize information on operating system, python version and dependencies.
```python
from mitiq import about
about()
```

Additional Python Environment Details (`pip freeze` or `conda list`):

Mitiq version 0.34
Qiskit 1.0.1

```
Copy and paste the output of `pip freeze` or `conda list` here.
```

</details>

### PR DESCRIPTION
**Pull request by andre-a-alves**
## Description

Closes #2208 #2204 #1422

Upgrades Qiskit to 1.0.2 and introduces a `TransformationPass` for transforming `QuantumRegister`s.

## ~Out of Scope~
~Docs have not been updated to Qiskit 1.0. Some examples from the docs use `qiskit.execute()` and remain incompatible with Qiskit 1.0.~

---

### License

- [X] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Before opening the PR, please ensure you have completed the following where appropriate.

- [X] I added unit tests for new code.
- [X] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [X] I used [Google-style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) docstrings for functions.
- [x] I [updated the documentation](../blob/main/docs/CONTRIBUTING_DOCS.md) where relevant.
- [x] Added myself / the copyright holder to the AUTHORS file

### CODE REVIEW

**Comment by andre-a-alves** [comment]
This morning I took a look at the failing job. The Python 3.9 failures all relate to a dependency conflict involving Sphinx, the Sphinx theme, and nbsphinx, which is unrelated to this PR. Later this week I can look into why the docs are failing. I did not update the examples that use `qiskit.execute()`, so that could be it, in which case I can update them in the troubleshooting.

**Comment by cosenal** [comment]
> Also, the 3.11 failure seems to be an issue with Codecov: `Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}`

This was resolved by #2271. If you rebase main into your branch, and restart the test build, it should be fixed.

**Comment by cosenal** [comment]
> This morning I took a look at the failing job. The Python 3.9 failures all relate to a dependency conflict involving Sphinx, the Sphinx theme, and nbsphinx, which is unrelated to this PR. Later this week I can look into why the docs are failing. I did not update the examples that use `qiskit.execute()`, so that could be it, in which case I can update them in the troubleshooting.

The docs build is failing, as you suspected, because the qiskit notebooks from the examples haven't been updated.

**Comment by andre-a-alves** [comment]
`main` was merged into the branch, and I updated the docs. The only remaining failures seem to be unrelated to this PR and have to do with incompatible depends in Python 3.9.

**Comment by cosenal** [comment]
Thanks, @andre-a-alves.

> The only remaining failures seem to be unrelated to this PR and have to do with incompatible depends in Python 3.9.

They are technically related to this PR in the sense that the conflicts are caused by the new dependencies introduced in this PR. In fact a clean build from the _main_ branch on Python 3.9 is still working. I am not sure why in this PR it's trying to install myst-parser 1.0.0 on Python 3.9 🤷‍♂️  I pushed a commit to pin myst-parser to >= 2.0.0, let's see if it works.

Also note the plan for Mitiq to drop support for Python 3.9 soon, see https://github.com/unitaryfund/mitiq/issues/2215.

**Comment by andre-a-alves** [comment]
I did not realize. If it doesn't pass with the myst-parser pin, this evening I'll look for a combination of dependencies that work.

File: mitiq/interface/mitiq_qiskit/transpiler.py
Comment (cosenal) [inline mitiq/interface/mitiq_qiskit/transpiler.py:13]: License-wise, it's all good. My concern about having this file in mitiq is that it's going to be hard to keep it synced with the upstream qiskit file. Can't we just use the imported functions as they are and wrap them when needed?

Comment (andre-a-alves) [reply mitiq/interface/mitiq_qiskit/transpiler.py:13]: The idea is not to keep them in sync. I used a lot of the original `ApplyLayout`, but I made substantive changes that are different from the original. The original class applies a `QuantumRegister` with the name `q`. We initialize the class with our own registers, and we use their names. From what I understand, maintaining the names is one of the purposes of the `_transform` function. But with probably half the class identical to the original, I thought it appropriate to include the license statement.

File: mitiq/interface/mitiq_qiskit/conversions.py
Comment (andre-a-alves) [reply mitiq/interface/mitiq_qiskit/conversions.py:217]: Good catch. I have modified the docstring and added a unit test

**Review by cosenal — APPROVED** [review]
LGTM 🚀 
Since it's quite a big change, let's wait for a second review before merging.

**Comment by nathanshammah** [comment]
Nice work @andre-a-alves!!

File: docs/source/examples/ibmq-backends.md
Comment (natestemen) [inline docs/source/examples/ibmq-backends.md:23]: Should we remove this line now?

File: docs/source/examples/pennylane-ibmq-backends.md
Comment (natestemen) [inline docs/source/examples/pennylane-ibmq-backends.md:65]: Should we add a `TODO` or something similar here so we can find it easily later? Otherwise I fear this comment will live on indefinitely 😛.

File: mitiq/interface/mitiq_qiskit/qiskit_utils.py
Comment (natestemen) [inline mitiq/interface/mitiq_qiskit/qiskit_utils.py:119]: I think the comment should go directly above `optimization_level` for clarity.

File: mitiq/interface/mitiq_qiskit/tests/test_transpiler.py
Comment (natestemen) [inline mitiq/interface/mitiq_qiskit/tests/test_transpiler.py:6]: Should it be this file?
```suggestion
"""Unit tests for qiskit TransformationPass (transpiler.py)."""
```

File: mitiq/interface/mitiq_qiskit/transpiler.py
Comment (natestemen) [inline mitiq/interface/mitiq_qiskit/transpiler.py:92]: Probably want to remove this `print` statement?

File: mitiq/interface/mitiq_qiskit/transpiler.py
Comment (natestemen) [inline mitiq/interface/mitiq_qiskit/transpiler.py:114]: If we are deleting a key-val pair from a dictionary if the key exists, we can do it with one lookup:

```suggestion
        self.property_set.pop("layout", None)
```

If this is less clear/readable, however, no need to accept.

**Review by natestemen — COMMENTED** [review]
Awesome work Andre! Thanks for the help on this 💪🏻.

I've left a few small comments, but all the major stuff is there. Qiskit users will be very grateful!

File: mitiq/interface/mitiq_qiskit/qiskit_utils.py
Comment (andre-a-alves) [reply mitiq/interface/mitiq_qiskit/qiskit_utils.py]: I went ahead and did the same thing in a few other places

File: docs/source/examples/pennylane-ibmq-backends.md
Comment (andre-a-alves) [reply docs/source/examples/pennylane-ibmq-backends.md:66]: TODO added

**Comment by andre-a-alves** [comment]
> Awesome work Andre! Thanks for the help on this 💪🏻.
> 
> I've left a few small comments, but all the major stuff is there. Qiskit users will be very grateful!

All comments addressed in latest commit. Good catches! Thank you for the review!

**Review by natestemen — APPROVED** [review]
:shipit:

**Pull request merged**
